src/ProbeBase.jl

```diff
@@ -2,138 +2,337 @@ module ProbeBase
-        end)[category] = nothing
     end
 end
-macro probe(kind::QuoteNode, category::QuoteNode, args=Expr(:tuple))
-    argspec = parse_args(args)
-    spec = ProbeSpec(__source__, kind.value, argspec.argtypes)
-    register_probe(__module__, category.value, spec)
-    :(probe_maybe_trigger($spec, $kind, $(argspec.argvalues...)))
+macro tracepoint(kind::QuoteNode, category::QuoteNode, lib_id, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, kind.value, lib_id, args)
+end
+macro region_start(category::QuoteNode, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, :start, 0, args)
 end
-#= FIXME
-macro probe_start(category::Symbol)
-    :(probe($spec, :start))
+macro region_finish(category::QuoteNode, lib_id, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, :finish, lib_id, args)
 end
-macro probe_stop(category::Symbol)
-    :(probe($spec, :stop))
+
+"""
+    @region :category arg1 arg2... ex
+
+Constructs a pair of tracepoints around the execution of `ex`. The start
+tracepoint executes just before `ex`, and the finish tracepoint executes just
+after `ex`. The category for both tracepoints is `:category`, and must be a
+literal `Symbol` (as the tracepoints are constructed while `@region` is being
+parsed). The kind for each tracepoint is `:start` and `:finish` respectively
+(used when programming or enabling/disabling tracepoints).
+
+These tracepoints start out disabled and unprogrammed, and by default will do
+nothing (having negligible performance impact, if any). They can be programmed
+with [`set!`](@ref), and enabled with [`enable!`](@ref).
+
+Arguments may be provided in a few formats:
+
+`arg::T` - This specifies that the argument with name `arg` has type `T`, where
+`T` is a concrete type. In this form, `arg` must be a variable in scope that
+`@region` can read during start and finish tracepoint execution. If `arg` is
+modified during the execution of `ex`, its new value will be passed into the
+finish tracepoint.
+
+`arg::T=value` - This format is similar to `arg::T`, except that `arg` is not
+read from a variable with the same name, but is instead hard-coded as `value`
+for both start and finish tracepoints.
+
+`(start_arg::ST, finish_arg::FT)` - This format is like `arg::T`, except that
+it allows specifying different values for the start and finish tracepoints.
+"""
+macro region(category, _args...)
+    if !(category isa QuoteNode)
+        throw(ArgumentError("@region: category must be a literal `Symbol`"))
+    end
+    args = Expr(:tuple, _args[1:end-1]...)
+    ex = _args[end]
+    start_args, finish_args = parse_region_args(args)
+    @gensym lib_id
+    quote
+        $lib_id = $(create_tracepoint!(__module__, __source__, category.value, :start, 0, start_args))
+        try
+            $(esc(ex))
+        finally
+            $(create_tracepoint!(__module__, __source__, category.value, :finish, lib_id, finish_args))
```

Comment:
Does this mean that the start and finish technically get their own enable bit?

Tracy requires that the `:finish` runs if-and-only-if the corresponding `:start` ran, and I don't think that's guaranteed here.

Reply:
Yes, that's true. The `enable`/`disable` code has methods that make it easy to ensure both get enabled or disabled (but it is certainly possible to enable only one or the other). We can document that manually enabling individual tracepoints is not recommended.

Reply:
Alternatively, we can hoist the semaphore and probe read within `@region` manually. I'll look into doing that.